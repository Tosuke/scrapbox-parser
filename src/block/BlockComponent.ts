export type BlockComponentType = {
  indent: number
  text: string
}

export const convertToBlockComponents = (blocks: string): Array<BlockComponentType> => (
  blocks.split('\n')
    .map((block: string): BlockComponentType => {
      const blockMatcher = block.match(/^\s*/)
      const indent = blockMatcher![0].length
      return { indent, text: block }
    })
)
