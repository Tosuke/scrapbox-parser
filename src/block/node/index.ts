import { QuoteNodeType, QuoteNodeParser } from './QuoteNode'
import { StrongImageNodeType, StrongImageNodeParser } from './StrongImageNode'
import { StrongIconNodeType, StrongIconNodeParser } from './StrongIconNode'
import { StrongNodeType, StrongNodeParser } from './StrongNode'
import { FormulaNodeType, FormulaNodeParser } from './FormulaNode'
import { DecorationNodeType, DecorationNodeParser } from './DecorationNode'
import { CodeNodeType, CodeNodeParser } from './CodeNode'
import { BlankNodeType, BlankNodeParser } from './BlankNode'
import { UrlNodeType, UrlNodeParser } from './UrlNode'
import { GoogleMapNodeType, GoogleMapNodeParser } from './GoogleMapNode'
import { InternalLinkNodeType, InternalLinkNodeParser } from './InternalLinkNode'
import { IconNodeType, IconNodeParser } from './IconNode'
import { HashTagNodeType, HashTagNodeParser } from './HashTagNode'
import { PlainNodeType, PlainNodeParser } from './PlainNode'

export type LineNodeType = QuoteNodeType
                         | StrongImageNodeType
                         | StrongIconNodeType
                         | StrongNodeType
                         | FormulaNodeType
                         | DecorationNodeType
                         | CodeNodeType
                         | BlankNodeType
                         | UrlNodeType
                         | GoogleMapNodeType
                         | InternalLinkNodeType
                         | IconNodeType
                         | HashTagNodeType
                         | PlainNodeType

export type ParserOptionType = {
  nested: boolean
  quoted: boolean
}
export type NextParserType = () => LineNodeType[]
export type ParserType = (text: string, opt: ParserOptionType, next: NextParserType) => LineNodeType[]

const FalsyEliminator: ParserType = (text, _opt, next) => {
  if (!text) return []
  return next()
}

const combineNodeParsers = (...parsers: ParserType[]) => {
  return (text: string = '', opt: ParserOptionType = { nested: false, quoted: false }): LineNodeType[] => (
    parsers.slice().reverse().reduce(
      (acc: NextParserType, parser: ParserType): NextParserType => () => parser(text, opt, acc),
      () => PlainNodeParser(text)
    )()
  )
}

export const convertToLineNodes = combineNodeParsers(
  FalsyEliminator,
  QuoteNodeParser,
  CodeNodeParser,
  FormulaNodeParser,
  BlankNodeParser,
  StrongImageNodeParser,
  StrongIconNodeParser,
  StrongNodeParser,
  UrlNodeParser,
  DecorationNodeParser,
  IconNodeParser,
  GoogleMapNodeParser,
  InternalLinkNodeParser,
  HashTagNodeParser
)
