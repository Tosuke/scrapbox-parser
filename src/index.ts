import { PageType, LineType, NodeType } from './types'
import parseToNodes from './parseToNodes'

const parse = (input: string): PageType => {
  const _lines: Array<string> = input.trim().split('\n')
  const title: string = _lines.shift() || 'Untitled'
  const lines: LineType[] = []

  for (const line of _lines) {
    const indentMatcher = line.match(/^\s*/)
    const indent: number = indentMatcher ? indentMatcher[0].length : 0
    const nodes: Array<NodeType> = parseToNodes(line.trim(), indent)
    lines.push({ indent, nodes })
  }

  return { title, lines }
}

export default parse
