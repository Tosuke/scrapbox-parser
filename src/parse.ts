import { LineType, LineComponentType, convertToLineComponents } from './line'
import parseToLines from './parseToLines'

type PageType = {
  title: string
  lines: Array<LineType>
}

const parse = (input: string): PageType => {
  const lineComponents: Array<LineComponentType> = convertToLineComponents(input.trim())

  const firstLine: LineComponentType = lineComponents.shift() || { indent: 0, text: '' }
  const title: string = firstLine.text || 'Untitled'
  const lines: Array<LineType> = parseToLines(lineComponents)

  return { title, lines }
}

export default parse
