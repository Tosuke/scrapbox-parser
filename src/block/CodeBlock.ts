import { BlockComponentType } from './BlockComponent'
import { PackedBlockComponentType } from './PackedBlockComponent'

export type CodeBlockComponentType = {
  type: 'codeBlock'
  components: Array<BlockComponentType>
}

export type CodeBlockType = {
  type: 'codeBlock'
  fileName: string
  content: string
}

export const isCodeBlockComponent = (packedBlockComponent: PackedBlockComponentType): packedBlockComponent is CodeBlockComponentType => (
  packedBlockComponent.type === 'codeBlock'
)

export const convertToCodeBlock = (blockComponent: CodeBlockComponentType): CodeBlockType => {
  const { components } = blockComponent
  const head: BlockComponentType = components.shift() || { indent: 0, text: '' }
  const { indent, text } = head
  const match = text.match(/^\s*code:(.+)$/)
  if (!match) {
    return {
      type: 'codeBlock',
      fileName: '',
      content: ''
    }
  }

  const fileName: string = match[1]
  return {
    type: 'codeBlock',
    fileName,
    content: components
      .map((blockComponent: BlockComponentType): string => blockComponent.text.substring(indent + 1))
      .join('\n')
  }
}
