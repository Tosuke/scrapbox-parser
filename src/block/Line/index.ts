import { BlockComponentType } from '../BlockComponent'
import { PackedBlockComponentType } from '../PackedBlockComponent'
import { createPlainNode } from './PlainNode'

export type LineComponentType = {
  type: 'line'
  component: BlockComponentType
}

export type LineNodeType = {}

export type LineType = {
  type: 'line'
  nodes: Array<LineNodeType>
}

export const isLineComponent = (packedBlockComponent: PackedBlockComponentType): packedBlockComponent is LineComponentType => (
  packedBlockComponent.type === 'line'
)

export const convertToLine = (lineComponent: LineComponentType): LineType => {
  return {
    type: 'line',
    nodes: [
      createPlainNode(lineComponent.component.text)
    ]
  }
}
