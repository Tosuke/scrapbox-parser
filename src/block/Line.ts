import { BlockComponentType } from './BlockComponent'
import { LineNodeType, convertToLineNodes } from './node'

export type LineComponentType = {
  type: 'line'
  component: BlockComponentType
}

export type LineType = {
  indent: number
  type: 'line'
  nodes: LineNodeType[]
}

export const convertToLine = (lineComponent: LineComponentType): LineType => {
  const { indent, text } = lineComponent.component
  return {
    indent,
    type: 'line',
    nodes: convertToLineNodes(text.substring(indent))
  }
}
