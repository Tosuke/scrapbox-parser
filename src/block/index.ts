import { PackedBlockComponentType } from './PackedBlockComponent'
import { TitleType, isTitleComponent, convertToTitle } from './Title'
import { CodeBlockType, isCodeBlockComponent, convertToCodeBlock } from './CodeBlock'
import { TableType, isTableComponent, convertToTable } from './Table'
import { LineType, convertToLine } from './Line'

export type BlockType = TitleType | CodeBlockType | TableType | LineType

export const convertToBlock = (packedBlockComponent: PackedBlockComponentType): BlockType => {
  if (isTitleComponent(packedBlockComponent)) return convertToTitle(packedBlockComponent)
  if (isCodeBlockComponent(packedBlockComponent)) return convertToCodeBlock(packedBlockComponent)
  if (isTableComponent(packedBlockComponent)) return convertToTable(packedBlockComponent)
  return convertToLine(packedBlockComponent)
}
