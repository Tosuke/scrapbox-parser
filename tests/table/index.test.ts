/* global describe it expect */
/* eslint-disable no-tabs, no-irregular-whitespace */

import { BlockComponentType, convertToBlockComponents } from '../../src/block/BlockComponent'
import { BlockType } from '../../src/block'
import { convertToBlocks } from '../../src/parse'

const defaultTableCells = [
  [
    [
      {
        type: 'plain',
        text: '1'
      }
    ],
    [
      {
        type: 'plain',
        text: '2'
      }
    ],
    [
      {
        type: 'plain',
        text: '3'
      }
    ]
  ],
  [
    [{
      type: 'plain',
      text: '1 '
    }],
    [
      {
        type: 'plain',
        text: '2 '
      }
    ],
    [
      {
        type: 'plain',
        text: '3'
      }
    ]
  ],
  [
    [
      {
        type: 'plain',
        text: '------'
      }
    ],
    [
      {
        type: 'plain',
        text: '------'
      }
    ],
    [
      {
        type: 'plain',
        text: '------'
      }
    ]
  ],
  [
    [
      {
        type: 'plain',
        text: 'a'
      }
    ],
    [
      {
        type: 'plain',
        text: 'b'
      }
    ],
    [
      {
        type: 'plain',
        text: 'c'
      }
    ]
  ]
]

describe('Table', () => {
  it('Simple table', () => {
    const input = `table:hello
${'\t'}1${'\t'}2${'\t'}3
${'\t'}1 ${'\t'}2 ${'\t'}3
${'\t'}------${'\t'}------${'\t'}------
${'\t'}a${'\t'}b${'\t'}c`
    const blockComponents: BlockComponentType[] = convertToBlockComponents(input)
    const blocks: BlockType[] = convertToBlocks(blockComponents)
    expect(blocks).toEqual([
      {
        indent: 0,
        type: 'table',
        fileName: 'hello',
        cells: defaultTableCells
      }
    ])
  })

  it('Bulleted table', () => {
    const input = ` table:bulleted
 ${'\t'}1${'\t'}2${'\t'}3
 ${'\t'}1 ${'\t'}2 ${'\t'}3
 ${'\t'}------${'\t'}------${'\t'}------
 ${'\t'}a${'\t'}b${'\t'}c`
    const blockComponents: BlockComponentType[] = convertToBlockComponents(input)
    const blocks: BlockType[] = convertToBlocks(blockComponents)
    expect(blocks).toEqual([
      {
        indent: 1,
        type: 'table',
        fileName: 'bulleted',
        cells: defaultTableCells
      }
    ])
  })

  it('Table with empty cells', () => {
    const input = `table:${' '}
${'\t'} ${'\t'}　${'\t'}${'  '}
${'\t'}${'\t'}${'\t'}`
    const blockComponents: BlockComponentType[] = convertToBlockComponents(input)
    const blocks: BlockType[] = convertToBlocks(blockComponents)
    expect(blocks).toEqual([
      {
        indent: 0,
        type: 'table',
        fileName: ' ',
        cells: [
          [
            [
              {
                type: 'plain',
                text: ' '
              }
            ],
            [
              {
                type: 'plain',
                text: '　'
              }
            ],
            [
              {
                type: 'plain',
                text: '  '
              }
            ]
          ],
          [
            [],
            [],
            []
          ]
        ]
      }
    ])
  })

  it('Staggered table', () => {
    const input = `table:Staggered
${'\t'}1${'\t'}2${'\t'}3${'\t'}4
${'\t'}1${'\t'}2${'\t'}3
${'\t'}1
${'\t'}1${'\t'}2
${'\t'}`
    const blockComponents: BlockComponentType[] = convertToBlockComponents(input)
    const blocks: BlockType[] = convertToBlocks(blockComponents)
    expect(blocks).toEqual([
      {
        indent: 0,
        type: 'table',
        fileName: 'Staggered',
        cells: [
          [
            [
              {
                type: 'plain',
                text: '1'
              }
            ],
            [
              {
                type: 'plain',
                text: '2'
              }
            ],
            [
              {
                type: 'plain',
                text: '3'
              }
            ],
            [
              {
                type: 'plain',
                text: '4'
              }
            ]
          ],
          [
            [
              {
                type: 'plain',
                text: '1'
              }
            ],
            [
              {
                type: 'plain',
                text: '2'
              }
            ],
            [
              {
                type: 'plain',
                text: '3'
              }
            ]
          ],
          [
            [
              {
                type: 'plain',
                text: '1'
              }
            ]
          ],
          [
            [
              {
                type: 'plain',
                text: '1'
              }
            ],
            [
              {
                type: 'plain',
                text: '2'
              }
            ]
          ],
          [
            []
          ]
        ]
      }
    ])
  })

  it('Consecutive table', () => {
    const input = `table:hello
${'\t'}1${'\t'}2${'\t'}3
${'\t'}1 ${'\t'}2 ${'\t'}3
${'\t'}------${'\t'}------${'\t'}------
${'\t'}a${'\t'}b${'\t'}c
table:hello
${'\t'}1${'\t'}2${'\t'}3
${'\t'}1 ${'\t'}2 ${'\t'}3
${'\t'}------${'\t'}------${'\t'}------
${'\t'}a${'\t'}b${'\t'}c`
    const blockComponents: BlockComponentType[] = convertToBlockComponents(input)
    const blocks: BlockType[] = convertToBlocks(blockComponents)
    expect(blocks).toEqual([
      {
        indent: 0,
        type: 'table',
        fileName: 'hello',
        cells: defaultTableCells
      },
      {
        indent: 0,
        type: 'table',
        fileName: 'hello',
        cells: defaultTableCells
      }
    ])
  })

  it('Table with link', () => {
    const input = `table:table with link
${'\t'}[Link]${'\t'}This is [Link]`
    const blockComponents: BlockComponentType[] = convertToBlockComponents(input)
    const blocks: BlockType[] = convertToBlocks(blockComponents)
    expect(blocks).toEqual([
      {
        indent: 0,
        type: 'table',
        fileName: 'table with link',
        cells: [
          [
            [
              {
                type: 'link',
                pathType: 'relative',
                href: 'Link'
              }
            ],
            [
              {
                type: 'plain',
                text: 'This is '
              },
              {
                type: 'link',
                pathType: 'relative',
                href: 'Link'
              }
            ]
          ]
        ]
      }
    ])
  })
})
