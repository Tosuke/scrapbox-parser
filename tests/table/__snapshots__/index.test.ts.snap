// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Bulleted table: toMatchSnapshotWhenParsing 1`] = `
Array [
  Object {
    "cells": Array [
      Array [
        Array [
          Object {
            "text": "1",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "2",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "3",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "text": "1 ",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "2 ",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "3",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "text": "------",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "------",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "------",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "text": "a",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "b",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "c",
            "type": "plain",
          },
        ],
      ],
    ],
    "fileName": "bulleted",
    "indent": 1,
    "type": "table",
  },
]
`;

exports[`Table Consecutive table: toMatchSnapshotWhenParsing 1`] = `
Array [
  Object {
    "cells": Array [
      Array [
        Array [
          Object {
            "text": "1",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "2",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "3",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "text": "1 ",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "2 ",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "3",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "text": "------",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "------",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "------",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "text": "a",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "b",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "c",
            "type": "plain",
          },
        ],
      ],
    ],
    "fileName": "hello",
    "indent": 0,
    "type": "table",
  },
  Object {
    "cells": Array [
      Array [
        Array [
          Object {
            "text": "1",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "2",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "3",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "text": "1 ",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "2 ",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "3",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "text": "------",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "------",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "------",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "text": "a",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "b",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "c",
            "type": "plain",
          },
        ],
      ],
    ],
    "fileName": "hello",
    "indent": 0,
    "type": "table",
  },
]
`;

exports[`Table Simple table: toMatchSnapshotWhenParsing 1`] = `
Array [
  Object {
    "cells": Array [
      Array [
        Array [
          Object {
            "text": "1",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "2",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "3",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "text": "1 ",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "2 ",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "3",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "text": "------",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "------",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "------",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "text": "a",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "b",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "c",
            "type": "plain",
          },
        ],
      ],
    ],
    "fileName": "hello",
    "indent": 0,
    "type": "table",
  },
]
`;

exports[`Table Staggered table: toMatchSnapshotWhenParsing 1`] = `
Array [
  Object {
    "cells": Array [
      Array [
        Array [
          Object {
            "text": "1",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "2",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "3",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "4",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "text": "1",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "2",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "3",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "text": "1",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "text": "1",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "2",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [],
      ],
    ],
    "fileName": "Staggered",
    "indent": 0,
    "type": "table",
  },
]
`;

exports[`Table Table with empty cells: toMatchSnapshotWhenParsing 1`] = `
Array [
  Object {
    "cells": Array [
      Array [
        Array [
          Object {
            "text": " ",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "ã€€",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "text": "  ",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [],
        Array [],
        Array [],
      ],
    ],
    "fileName": " ",
    "indent": 0,
    "type": "table",
  },
]
`;

exports[`Table Table with link: toMatchSnapshotWhenParsing 1`] = `
Array [
  Object {
    "cells": Array [
      Array [
        Array [
          Object {
            "content": "",
            "href": "Link",
            "pathType": "relative",
            "type": "link",
          },
        ],
        Array [
          Object {
            "text": "This is ",
            "type": "plain",
          },
          Object {
            "content": "",
            "href": "Link",
            "pathType": "relative",
            "type": "link",
          },
        ],
      ],
    ],
    "fileName": "table with link",
    "indent": 0,
    "type": "table",
  },
]
`;
